File: rb/IpinfoDeveloper_sdk.rb
# IpinfoDeveloper SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'IpinfoDeveloper_types'


class IpinfoDeveloperSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = IpinfoDeveloperUtility.new
    @_utility = utility

    config = IpinfoDeveloperConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = IpinfoDeveloperHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = IpinfoDeveloperHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, IpinfoDeveloperFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    IpinfoDeveloperUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = IpinfoDeveloperHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = IpinfoDeveloperHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = IpinfoDeveloperHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = IpinfoDeveloperSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    utility.make_fetch_def.call(ctx)
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue IpinfoDeveloperError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = IpinfoDeveloperHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = IpinfoDeveloperHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Idiomatic facade: client.abuse.list / client.abuse.load({ "id" => ... })
  def abuse
    require_relative 'entity/abuse_entity'
    @abuse ||= AbuseEntity.new(self, nil)
  end

  # Deprecated: use client.abuse instead.
  def Abuse(data = nil)
    require_relative 'entity/abuse_entity'
    AbuseEntity.new(self, data)
  end


  # Idiomatic facade: client.asn.list / client.asn.load({ "id" => ... })
  def asn
    require_relative 'entity/asn_entity'
    @asn ||= AsnEntity.new(self, nil)
  end

  # Deprecated: use client.asn instead.
  def Asn(data = nil)
    require_relative 'entity/asn_entity'
    AsnEntity.new(self, data)
  end


  # Idiomatic facade: client.carrier.list / client.carrier.load({ "id" => ... })
  def carrier
    require_relative 'entity/carrier_entity'
    @carrier ||= CarrierEntity.new(self, nil)
  end

  # Deprecated: use client.carrier instead.
  def Carrier(data = nil)
    require_relative 'entity/carrier_entity'
    CarrierEntity.new(self, data)
  end


  # Idiomatic facade: client.company.list / client.company.load({ "id" => ... })
  def company
    require_relative 'entity/company_entity'
    @company ||= CompanyEntity.new(self, nil)
  end

  # Deprecated: use client.company instead.
  def Company(data = nil)
    require_relative 'entity/company_entity'
    CompanyEntity.new(self, data)
  end


  # Idiomatic facade: client.core.list / client.core.load({ "id" => ... })
  def core
    require_relative 'entity/core_entity'
    @core ||= CoreEntity.new(self, nil)
  end

  # Deprecated: use client.core instead.
  def Core(data = nil)
    require_relative 'entity/core_entity'
    CoreEntity.new(self, data)
  end


  # Idiomatic facade: client.domain.list / client.domain.load({ "id" => ... })
  def domain
    require_relative 'entity/domain_entity'
    @domain ||= DomainEntity.new(self, nil)
  end

  # Deprecated: use client.domain instead.
  def Domain(data = nil)
    require_relative 'entity/domain_entity'
    DomainEntity.new(self, data)
  end


  # Idiomatic facade: client.general.list / client.general.load({ "id" => ... })
  def general
    require_relative 'entity/general_entity'
    @general ||= GeneralEntity.new(self, nil)
  end

  # Deprecated: use client.general instead.
  def General(data = nil)
    require_relative 'entity/general_entity'
    GeneralEntity.new(self, data)
  end


  # Idiomatic facade: client.get_current_information.list / client.get_current_information.load({ "id" => ... })
  def get_current_information
    require_relative 'entity/get_current_information_entity'
    @get_current_information ||= GetCurrentInformationEntity.new(self, nil)
  end

  # Deprecated: use client.get_current_information instead.
  def GetCurrentInformation(data = nil)
    require_relative 'entity/get_current_information_entity'
    GetCurrentInformationEntity.new(self, data)
  end


  # Idiomatic facade: client.get_information_by_ip.list / client.get_information_by_ip.load({ "id" => ... })
  def get_information_by_ip
    require_relative 'entity/get_information_by_ip_entity'
    @get_information_by_ip ||= GetInformationByIpEntity.new(self, nil)
  end

  # Deprecated: use client.get_information_by_ip instead.
  def GetInformationByIp(data = nil)
    require_relative 'entity/get_information_by_ip_entity'
    GetInformationByIpEntity.new(self, data)
  end


  # Idiomatic facade: client.ipinfo_core.list / client.ipinfo_core.load({ "id" => ... })
  def ipinfo_core
    require_relative 'entity/ipinfo_core_entity'
    @ipinfo_core ||= IpinfoCoreEntity.new(self, nil)
  end

  # Deprecated: use client.ipinfo_core instead.
  def IpinfoCore(data = nil)
    require_relative 'entity/ipinfo_core_entity'
    IpinfoCoreEntity.new(self, data)
  end


  # Idiomatic facade: client.ipinfo_lite.list / client.ipinfo_lite.load({ "id" => ... })
  def ipinfo_lite
    require_relative 'entity/ipinfo_lite_entity'
    @ipinfo_lite ||= IpinfoLiteEntity.new(self, nil)
  end

  # Deprecated: use client.ipinfo_lite instead.
  def IpinfoLite(data = nil)
    require_relative 'entity/ipinfo_lite_entity'
    IpinfoLiteEntity.new(self, data)
  end


  # Idiomatic facade: client.ipinfo_plus.list / client.ipinfo_plus.load({ "id" => ... })
  def ipinfo_plus
    require_relative 'entity/ipinfo_plus_entity'
    @ipinfo_plus ||= IpinfoPlusEntity.new(self, nil)
  end

  # Deprecated: use client.ipinfo_plus instead.
  def IpinfoPlus(data = nil)
    require_relative 'entity/ipinfo_plus_entity'
    IpinfoPlusEntity.new(self, data)
  end


  # Idiomatic facade: client.lite.list / client.lite.load({ "id" => ... })
  def lite
    require_relative 'entity/lite_entity'
    @lite ||= LiteEntity.new(self, nil)
  end

  # Deprecated: use client.lite instead.
  def Lite(data = nil)
    require_relative 'entity/lite_entity'
    LiteEntity.new(self, data)
  end


  # Idiomatic facade: client.max.list / client.max.load({ "id" => ... })
  def max
    require_relative 'entity/max_entity'
    @max ||= MaxEntity.new(self, nil)
  end

  # Deprecated: use client.max instead.
  def Max(data = nil)
    require_relative 'entity/max_entity'
    MaxEntity.new(self, data)
  end


  # Idiomatic facade: client.men.list / client.men.load({ "id" => ... })
  def men
    require_relative 'entity/men_entity'
    @men ||= MenEntity.new(self, nil)
  end

  # Deprecated: use client.men instead.
  def Men(data = nil)
    require_relative 'entity/men_entity'
    MenEntity.new(self, data)
  end


  # Idiomatic facade: client.place.list / client.place.load({ "id" => ... })
  def place
    require_relative 'entity/place_entity'
    @place ||= PlaceEntity.new(self, nil)
  end

  # Deprecated: use client.place instead.
  def Place(data = nil)
    require_relative 'entity/place_entity'
    PlaceEntity.new(self, data)
  end


  # Idiomatic facade: client.plus.list / client.plus.load({ "id" => ... })
  def plus
    require_relative 'entity/plus_entity'
    @plus ||= PlusEntity.new(self, nil)
  end

  # Deprecated: use client.plus instead.
  def Plus(data = nil)
    require_relative 'entity/plus_entity'
    PlusEntity.new(self, data)
  end


  # Idiomatic facade: client.privacy.list / client.privacy.load({ "id" => ... })
  def privacy
    require_relative 'entity/privacy_entity'
    @privacy ||= PrivacyEntity.new(self, nil)
  end

  # Deprecated: use client.privacy instead.
  def Privacy(data = nil)
    require_relative 'entity/privacy_entity'
    PrivacyEntity.new(self, data)
  end


  # Idiomatic facade: client.privacy_extended.list / client.privacy_extended.load({ "id" => ... })
  def privacy_extended
    require_relative 'entity/privacy_extended_entity'
    @privacy_extended ||= PrivacyExtendedEntity.new(self, nil)
  end

  # Deprecated: use client.privacy_extended instead.
  def PrivacyExtended(data = nil)
    require_relative 'entity/privacy_extended_entity'
    PrivacyExtendedEntity.new(self, data)
  end


  # Idiomatic facade: client.range.list / client.range.load({ "id" => ... })
  def range
    require_relative 'entity/range_entity'
    @range ||= RangeEntity.new(self, nil)
  end

  # Deprecated: use client.range instead.
  def Range(data = nil)
    require_relative 'entity/range_entity'
    RangeEntity.new(self, data)
  end


  # Idiomatic facade: client.residential_proxy.list / client.residential_proxy.load({ "id" => ... })
  def residential_proxy
    require_relative 'entity/residential_proxy_entity'
    @residential_proxy ||= ResidentialProxyEntity.new(self, nil)
  end

  # Deprecated: use client.residential_proxy instead.
  def ResidentialProxy(data = nil)
    require_relative 'entity/residential_proxy_entity'
    ResidentialProxyEntity.new(self, data)
  end


  # Idiomatic facade: client.single.list / client.single.load({ "id" => ... })
  def single
    require_relative 'entity/single_entity'
    @single ||= SingleEntity.new(self, nil)
  end

  # Deprecated: use client.single instead.
  def Single(data = nil)
    require_relative 'entity/single_entity'
    SingleEntity.new(self, data)
  end


  # Idiomatic facade: client.whois_asn.list / client.whois_asn.load({ "id" => ... })
  def whois_asn
    require_relative 'entity/whois_asn_entity'
    @whois_asn ||= WhoisAsnEntity.new(self, nil)
  end

  # Deprecated: use client.whois_asn instead.
  def WhoisAsn(data = nil)
    require_relative 'entity/whois_asn_entity'
    WhoisAsnEntity.new(self, data)
  end


  # Idiomatic facade: client.whois_domain.list / client.whois_domain.load({ "id" => ... })
  def whois_domain
    require_relative 'entity/whois_domain_entity'
    @whois_domain ||= WhoisDomainEntity.new(self, nil)
  end

  # Deprecated: use client.whois_domain instead.
  def WhoisDomain(data = nil)
    require_relative 'entity/whois_domain_entity'
    WhoisDomainEntity.new(self, data)
  end


  # Idiomatic facade: client.whois_ip.list / client.whois_ip.load({ "id" => ... })
  def whois_ip
    require_relative 'entity/whois_ip_entity'
    @whois_ip ||= WhoisIpEntity.new(self, nil)
  end

  # Deprecated: use client.whois_ip instead.
  def WhoisIp(data = nil)
    require_relative 'entity/whois_ip_entity'
    WhoisIpEntity.new(self, data)
  end


  # Idiomatic facade: client.whois_net_id.list / client.whois_net_id.load({ "id" => ... })
  def whois_net_id
    require_relative 'entity/whois_net_id_entity'
    @whois_net_id ||= WhoisNetIdEntity.new(self, nil)
  end

  # Deprecated: use client.whois_net_id instead.
  def WhoisNetId(data = nil)
    require_relative 'entity/whois_net_id_entity'
    WhoisNetIdEntity.new(self, data)
  end


  # Idiomatic facade: client.whois_org.list / client.whois_org.load({ "id" => ... })
  def whois_org
    require_relative 'entity/whois_org_entity'
    @whois_org ||= WhoisOrgEntity.new(self, nil)
  end

  # Deprecated: use client.whois_org instead.
  def WhoisOrg(data = nil)
    require_relative 'entity/whois_org_entity'
    WhoisOrgEntity.new(self, data)
  end


  # Idiomatic facade: client.whois_poc.list / client.whois_poc.load({ "id" => ... })
  def whois_poc
    require_relative 'entity/whois_poc_entity'
    @whois_poc ||= WhoisPocEntity.new(self, nil)
  end

  # Deprecated: use client.whois_poc instead.
  def WhoisPoc(data = nil)
    require_relative 'entity/whois_poc_entity'
    WhoisPocEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = IpinfoDeveloperSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/ipinfodeveloper_sdk.py
# IpinfoDeveloper SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import IpinfoDeveloperUtility
from core.spec import IpinfoDeveloperSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import IpinfoDeveloperBaseFeature
from features import _make_feature


class IpinfoDeveloperSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = IpinfoDeveloperUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features in the resolved order (make_options puts an explicit
        # list order first, else defaults to test-first). Ordering matters: the
        # `test` feature installs the base mock transport and the transport
        # features (retry/cache/netsim/proxy/ratelimit) wrap whatever is
        # current, so `test` must be added before them to sit at the base.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            featureorder = vs.getpath(self.options, "__derived__.featureorder")
            if isinstance(featureorder, list):
                for fname in featureorder:
                    fopts = helpers.to_map(feature_opts.get(fname))
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return IpinfoDeveloperUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = IpinfoDeveloperSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def Abuse(self, data=None) -> "AbuseEntity":
        """Entity factory: client.Abuse().list() / client.Abuse().load({"id": ...})."""
        from entity.abuse_entity import AbuseEntity
        return AbuseEntity(self, data)


    def Asn(self, data=None) -> "AsnEntity":
        """Entity factory: client.Asn().list() / client.Asn().load({"id": ...})."""
        from entity.asn_entity import AsnEntity
        return AsnEntity(self, data)


    def Carrier(self, data=None) -> "CarrierEntity":
        """Entity factory: client.Carrier().list() / client.Carrier().load({"id": ...})."""
        from entity.carrier_entity import CarrierEntity
        return CarrierEntity(self, data)


    def Company(self, data=None) -> "CompanyEntity":
        """Entity factory: client.Company().list() / client.Company().load({"id": ...})."""
        from entity.company_entity import CompanyEntity
        return CompanyEntity(self, data)


    def Core(self, data=None) -> "CoreEntity":
        """Entity factory: client.Core().list() / client.Core().load({"id": ...})."""
        from entity.core_entity import CoreEntity
        return CoreEntity(self, data)


    def Domain(self, data=None) -> "DomainEntity":
        """Entity factory: client.Domain().list() / client.Domain().load({"id": ...})."""
        from entity.domain_entity import DomainEntity
        return DomainEntity(self, data)


    def General(self, data=None) -> "GeneralEntity":
        """Entity factory: client.General().list() / client.General().load({"id": ...})."""
        from entity.general_entity import GeneralEntity
        return GeneralEntity(self, data)


    def GetCurrentInformation(self, data=None) -> "GetCurrentInformationEntity":
        """Entity factory: client.GetCurrentInformation().list() / client.GetCurrentInformation().load({"id": ...})."""
        from entity.get_current_information_entity import GetCurrentInformationEntity
        return GetCurrentInformationEntity(self, data)


    def GetInformationByIp(self, data=None) -> "GetInformationByIpEntity":
        """Entity factory: client.GetInformationByIp().list() / client.GetInformationByIp().load({"id": ...})."""
        from entity.get_information_by_ip_entity import GetInformationByIpEntity
        return GetInformationByIpEntity(self, data)


    def IpinfoCore(self, data=None) -> "IpinfoCoreEntity":
        """Entity factory: client.IpinfoCore().list() / client.IpinfoCore().load({"id": ...})."""
        from entity.ipinfo_core_entity import IpinfoCoreEntity
        return IpinfoCoreEntity(self, data)


    def IpinfoLite(self, data=None) -> "IpinfoLiteEntity":
        """Entity factory: client.IpinfoLite().list() / client.IpinfoLite().load({"id": ...})."""
        from entity.ipinfo_lite_entity import IpinfoLiteEntity
        return IpinfoLiteEntity(self, data)


    def IpinfoPlus(self, data=None) -> "IpinfoPlusEntity":
        """Entity factory: client.IpinfoPlus().list() / client.IpinfoPlus().load({"id": ...})."""
        from entity.ipinfo_plus_entity import IpinfoPlusEntity
        return IpinfoPlusEntity(self, data)


    def Lite(self, data=None) -> "LiteEntity":
        """Entity factory: client.Lite().list() / client.Lite().load({"id": ...})."""
        from entity.lite_entity import LiteEntity
        return LiteEntity(self, data)


    def Max(self, data=None) -> "MaxEntity":
        """Entity factory: client.Max().list() / client.Max().load({"id": ...})."""
        from entity.max_entity import MaxEntity
        return MaxEntity(self, data)


    def Men(self, data=None) -> "MenEntity":
        """Entity factory: client.Men().list() / client.Men().load({"id": ...})."""
        from entity.men_entity import MenEntity
        return MenEntity(self, data)


    def Place(self, data=None) -> "PlaceEntity":
        """Entity factory: client.Place().list() / client.Place().load({"id": ...})."""
        from entity.place_entity import PlaceEntity
        return PlaceEntity(self, data)


    def Plus(self, data=None) -> "PlusEntity":
        """Entity factory: client.Plus().list() / client.Plus().load({"id": ...})."""
        from entity.plus_entity import PlusEntity
        return PlusEntity(self, data)


    def Privacy(self, data=None) -> "PrivacyEntity":
        """Entity factory: client.Privacy().list() / client.Privacy().load({"id": ...})."""
        from entity.privacy_entity import PrivacyEntity
        return PrivacyEntity(self, data)


    def PrivacyExtended(self, data=None) -> "PrivacyExtendedEntity":
        """Entity factory: client.PrivacyExtended().list() / client.PrivacyExtended().load({"id": ...})."""
        from entity.privacy_extended_entity import PrivacyExtendedEntity
        return PrivacyExtendedEntity(self, data)


    def Range(self, data=None) -> "RangeEntity":
        """Entity factory: client.Range().list() / client.Range().load({"id": ...})."""
        from entity.range_entity import RangeEntity
        return RangeEntity(self, data)


    def ResidentialProxy(self, data=None) -> "ResidentialProxyEntity":
        """Entity factory: client.ResidentialProxy().list() / client.ResidentialProxy().load({"id": ...})."""
        from entity.residential_proxy_entity import ResidentialProxyEntity
        return ResidentialProxyEntity(self, data)


    def Single(self, data=None) -> "SingleEntity":
        """Entity factory: client.Single().list() / client.Single().load({"id": ...})."""
        from entity.single_entity import SingleEntity
        return SingleEntity(self, data)


    def WhoisAsn(self, data=None) -> "WhoisAsnEntity":
        """Entity factory: client.WhoisAsn().list() / client.WhoisAsn().load({"id": ...})."""
        from entity.whois_asn_entity import WhoisAsnEntity
        return WhoisAsnEntity(self, data)


    def WhoisDomain(self, data=None) -> "WhoisDomainEntity":
        """Entity factory: client.WhoisDomain().list() / client.WhoisDomain().load({"id": ...})."""
        from entity.whois_domain_entity import WhoisDomainEntity
        return WhoisDomainEntity(self, data)


    def WhoisIp(self, data=None) -> "WhoisIpEntity":
        """Entity factory: client.WhoisIp().list() / client.WhoisIp().load({"id": ...})."""
        from entity.whois_ip_entity import WhoisIpEntity
        return WhoisIpEntity(self, data)


    def WhoisNetId(self, data=None) -> "WhoisNetIdEntity":
        """Entity factory: client.WhoisNetId().list() / client.WhoisNetId().load({"id": ...})."""
        from entity.whois_net_id_entity import WhoisNetIdEntity
        return WhoisNetIdEntity(self, data)


    def WhoisOrg(self, data=None) -> "WhoisOrgEntity":
        """Entity factory: client.WhoisOrg().list() / client.WhoisOrg().load({"id": ...})."""
        from entity.whois_org_entity import WhoisOrgEntity
        return WhoisOrgEntity(self, data)


    def WhoisPoc(self, data=None) -> "WhoisPocEntity":
        """Entity factory: client.WhoisPoc().list() / client.WhoisPoc().load({"id": ...})."""
        from entity.whois_poc_entity import WhoisPocEntity
        return WhoisPocEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "IpinfoDeveloperSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.abuse_entity import AbuseEntity
    from entity.asn_entity import AsnEntity
    from entity.carrier_entity import CarrierEntity
    from entity.company_entity import CompanyEntity
    from entity.core_entity import CoreEntity
    from entity.domain_entity import DomainEntity
    from entity.general_entity import GeneralEntity
    from entity.get_current_information_entity import GetCurrentInformationEntity
    from entity.get_information_by_ip_entity import GetInformationByIpEntity
    from entity.ipinfo_core_entity import IpinfoCoreEntity
    from entity.ipinfo_lite_entity import IpinfoLiteEntity
    from entity.ipinfo_plus_entity import IpinfoPlusEntity
    from entity.lite_entity import LiteEntity
    from entity.max_entity import MaxEntity
    from entity.men_entity import MenEntity
    from entity.place_entity import PlaceEntity
    from entity.plus_entity import PlusEntity
    from entity.privacy_entity import PrivacyEntity
    from entity.privacy_extended_entity import PrivacyExtendedEntity
    from entity.range_entity import RangeEntity
    from entity.residential_proxy_entity import ResidentialProxyEntity
    from entity.single_entity import SingleEntity
    from entity.whois_asn_entity import WhoisAsnEntity
    from entity.whois_domain_entity import WhoisDomainEntity
    from entity.whois_ip_entity import WhoisIpEntity
    from entity.whois_net_id_entity import WhoisNetIdEntity
    from entity.whois_org_entity import WhoisOrgEntity
    from entity.whois_poc_entity import WhoisPocEntity
